File: lib/src/main/java/com/github/preference/NumberPickerPreference.java
/*
 * Copyright 2012, Moshe Waisberg
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.preference;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.res.TypedArray;
import android.util.AttributeSet;

import androidx.annotation.Nullable;
import androidx.core.content.res.TypedArrayUtils;

import com.github.lib.R;

/**
 * Preference that shows a number picker.
 *
 * @author Moshe Waisberg
 */
public class NumberPickerPreference extends DialogPreference {

    private static final int[] ATTRIBUTES = {android.R.attr.max};

    private int value;
    private int max;
    private int min;
    private boolean progressSet;

    public NumberPickerPreference(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);

        final TypedArray a = context.obtainStyledAttributes(attrs, ATTRIBUTES, defStyleAttr, defStyleRes);
        this.max = a.getInt(0, 100);
        a.recycle();

        if (getDialogLayoutResource() == 0) {
            setDialogLayoutResource(R.layout.preference_dialog_numberpicker);
        }
    }

    public NumberPickerPreference(Context context, AttributeSet attrs, int defStyleAttr) {
        this(context, attrs, defStyleAttr, 0);
    }

    public NumberPickerPreference(Context context, AttributeSet attrs) {
        this(context, attrs, TypedArrayUtils.getAttr(context, R.attr.dialogPreferenceStyle, android.R.attr.dialogPreferenceStyle));
    }

    public NumberPickerPreference(Context context) {
        this(context, null);
    }

    @Override
    protected Integer onGetDefaultValue(TypedArray a, int index) {
        return a.hasValue(0) ? a.getInt(index, 0) : null;
    }

    @Override
    protected void onSetInitialValue(@Nullable Object defaultValue) {
        setValue(getPersistedInt((defaultValue != null) ? (Integer) defaultValue : value));
    }

    /**
     * Saves the value to the {@link SharedPreferences}.
     *
     * @param value the value.
     */
    public void setValue(int value) {
        // Always persist/notify the first time; don't assume the field's default value.
        final boolean changed = this.value != value;
        if (changed || !progressSet) {
            this.value = value;
            progressSet = true;
            persistInt(value);
            if (changed) {
                notifyDependencyChange(shouldDisableDependents());
                notifyChanged();
            }
        }
    }

    /**
     * Gets the value from the {@link SharedPreferences}.
     *
     * @return the value.
     */
    public int getValue() {
        return value;
    }

    /**
     * Set the range of the value bar to {@code 0}...{@code max}.
     *
     * @param max the upper range of this value bar.
     */
    public void setMax(int max) {
        this.max = max;
    }

    /**
     * Get the maximum value.
     *
     * @return the upper range of this value bar.
     */
    public int getMax() {
        return max;
    }

    /**
     * Set the range of the value bar to {@code min}...{@code max}.
     *
     * @param min the lower range of this value bar.
     */
    public void setMin(int min) {
        this.min = min;
    }

    /**
     * Get the minimum value.
     *
     * @return the lower range of this value bar.
     */
    public int getMin() {
        return min;
    }
}

File: lib/src/main/java/com/github/preference/NumberPickerPreference.kt
/*
 * Copyright 2012, Moshe Waisberg
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.preference

import android.content.Context
import android.content.res.TypedArray
import android.os.Build
import android.util.AttributeSet
import androidx.annotation.AttrRes
import androidx.annotation.StyleRes
import com.github.lib.R
import com.github.util.TypedValueUtils
import kotlin.math.max
import kotlin.math.min

/**
 * Preference that shows a number picker.
 *
 * @author Moshe Waisberg
 */
open class NumberPickerPreference @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    @AttrRes defStyleAttr: Int = TypedValueUtils.getAttr(
        context,
        androidx.preference.R.attr.dialogPreferenceStyle,
        android.R.attr.dialogPreferenceStyle
    ),
    @StyleRes defStyleRes: Int = 0
) : DialogPreference(context, attrs, defStyleAttr, defStyleRes) {
    /**
     * The value.
     */
    var value = 0
        set(value) {
            field = setValueImpl(field, value)
        }

    /**
     * The upper bound.
     */
    var max: Int = 100
        set(value) {
            val newValue = max(min, value)
            if (newValue != field) {
                field = newValue
                notifyChanged()
            }
        }

    /**
     * The lower bound.
     */
    var min = 0
        set(value) {
            val newValue = min(max, value)
            if (newValue != field) {
                field = newValue
                notifyChanged()
            }
        }

    var displayedValues: Array<String?>? = null
        protected set

    private var progressSet = false

    init {
        var a = context.obtainStyledAttributes(
            attrs,
            androidx.preference.R.styleable.ListPreference,
            defStyleAttr,
            defStyleRes
        )
        displayedValues = TypedValueUtils.getStringArray(
            a,
            androidx.preference.R.styleable.ListPreference_entryValues,
            androidx.preference.R.styleable.ListPreference_android_entryValues
        )
        a.recycle()

        a = context.obtainStyledAttributes(
            attrs,
            R.styleable.NumberPickerPreference,
            defStyleAttr,
            defStyleRes
        )
        min = max(0, a.getInt(R.styleable.NumberPickerPreference_min, min))
        max = max(0, a.getInt(R.styleable.NumberPickerPreference_android_max, max))
        if (a.getBoolean(R.styleable.NumberPickerPreference_useSimpleSummaryProvider, false)) {
            summaryProvider = SimpleSummaryProvider.instance
        }
        a.recycle()

        if (dialogLayoutResource == 0) {
            dialogLayoutResource = R.layout.preference_dialog_numberpicker
        }
    }

    override fun onGetDefaultValue(a: TypedArray, index: Int): Int? {
        return if (a.hasValue(0)) a.getInt(index, 0) else null
    }

    override fun onSetInitialValue(defaultValue: Any?) {
        value = getPersistedInt((defaultValue as? Int) ?: value)
    }

    /**
     * Saves the value to the [SharedPreferences].
     *
     * @param oldValue The old value.
     * @param newValue The new value.
     */
    protected open fun setValueImpl(oldValue: Int, newValue: Int): Int {
        // Always persist/notify the first time; don't assume the field's default value.
        val changed = oldValue != newValue
        if (changed || !progressSet) {
            progressSet = true
            persistInt(newValue)
            if (changed) {
                notifyDependencyChange(shouldDisableDependents())
                notifyChanged()
            }
        }
        return newValue
    }

    /**
     * A simple [androidx.preference.Preference.SummaryProvider] implementation for a
     * [NumberPickerPreference]. If no value has been set, the summary displayed will be 'Not set',
     * otherwise the summary displayed will be the entry set for this preference.
     */
    class SimpleSummaryProvider private constructor() :
        SummaryProvider<NumberPickerPreference> {
        override fun provideSummary(preference: NumberPickerPreference): CharSequence? {
            var index = preference.value
            if (index < 0) return null
            val entryValues = preference.displayedValues
            if (entryValues != null) {
                val lastIndex = entryValues.lastIndex
                index = min(lastIndex, index - preference.min)
                return entryValues[index]
            }
            return index.toString()
        }

        companion object {
            /**
             * Retrieve a singleton instance of this simple
             * [androidx.preference.Preference.SummaryProvider] implementation.
             */
            val instance: SimpleSummaryProvider by lazy { SimpleSummaryProvider() }
        }
    }

    companion object {
        private val attrAndroidMin = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            android.R.attr.min
        } else {
            androidx.preference.R.attr.min
        }
    }
}